// angular2gridster/esm2015/lib/gridster-prototype/gridster-item-prototype.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Input, Output, HostBinding, EventEmitter, OnInit, OnDestroy, NgZone } from '@angular/core';
import { fromEvent } from 'rxjs';
import { GridsterPrototypeService } from './gridster-prototype.service';
import { GridListItem } from '../gridList/GridListItem';
import { Draggable } from '../utils/draggable';
import { utils } from '../utils/utils';
let GridsterItemPrototypeDirective = class GridsterItemPrototypeDirective {
    constructor(zone, elementRef, gridsterPrototype) {
        this.zone = zone;
        this.elementRef = elementRef;
        this.gridsterPrototype = gridsterPrototype;
        this.drop = new EventEmitter();
        this.start = new EventEmitter();
        this.cancel = new EventEmitter();
        this.enter = new EventEmitter();
        this.out = new EventEmitter();
        this.config = {};
        this.x = 0;
        this.y = 0;
        this.autoSize = false;
        this.isDragging = false;
        this.subscribtions = [];
        this.item = (new GridListItem()).setFromGridsterItemPrototype(this);
    }
    // must be set to true because of item dragAndDrop configuration
    get dragAndDrop() {
        return true;
    }
    get gridster() {
        return this.dragContextGridster;
    }
    ngOnInit() {
        this.wSm = this.wSm || this.w;
        this.hSm = this.hSm || this.h;
        this.wMd = this.wMd || this.w;
        this.hMd = this.hMd || this.h;
        this.wLg = this.wLg || this.w;
        this.hLg = this.hLg || this.h;
        this.wXl = this.wXl || this.w;
        this.hXl = this.hXl || this.h;
        this.zone.runOutsideAngular(() => {
            this.enableDragDrop();
        });
    }
    ngOnDestroy() {
        this.subscribtions.forEach((sub) => {
            sub.unsubscribe();
        });
    }
    onDrop(gridster) {
        if (!this.config.helper) {
            this.$element.parentNode.removeChild(this.$element);
        }
        this.drop.emit({
            item: this.item,
            gridster: gridster
        });
    }
    onCancel() {
        this.cancel.emit({ item: this.item });
    }
    onEnter(gridster) {
        this.enter.emit({
            item: this.item,
            gridster: gridster
        });
    }
    onOver(gridster) { }
    onOut(gridster) {
        this.out.emit({
            item: this.item,
            gridster: gridster
        });
    }
    getPositionToGridster(gridster) {
        const relativeContainerCoords = this.getContainerCoordsToGridster(gridster);
        return {
            y: this.positionY - relativeContainerCoords.top,
            x: this.positionX - relativeContainerCoords.left
        };
    }
    setDragContextGridster(gridster) {
        this.dragContextGridster = gridster;
    }
    getContainerCoordsToGridster(gridster) {
        return {
            left: gridster.gridsterRect.left - this.parentRect.left,
            top: gridster.gridsterRect.top - this.parentRect.top
        };
    }
    enableDragDrop() {
        let cursorToElementPosition;
        const draggable = new Draggable(this.elementRef.nativeElement);
        const dragStartSub = draggable.dragStart
            .subscribe((event) => {
            this.zone.run(() => {
                this.$element = this.provideDragElement();
                this.containerRectange = this.$element.parentElement.getBoundingClientRect();
                this.updateParentElementData();
                this.onStart(event);
                cursorToElementPosition = event.getRelativeCoordinates(this.$element);
            });
        });
        const dragSub = draggable.dragMove
            .subscribe((event) => {
            this.setElementPosition(this.$element, {
                x: event.clientX - cursorToElementPosition.x - this.parentRect.left,
                y: event.clientY - cursorToElementPosition.y - this.parentRect.top
            });
            this.onDrag(event);
        });
        const dragStopSub = draggable.dragStop
            .subscribe((event) => {
            this.zone.run(() => {
                this.onStop(event);
                this.$element = null;
            });
        });
        const scrollSub = fromEvent(document, 'scroll')
            .subscribe(() => {
            if (this.$element) {
                this.updateParentElementData();
            }
        });
        this.subscribtions = this.subscribtions.concat([dragStartSub, dragSub, dragStopSub, scrollSub]);
    }
    setElementPosition(element, position) {
        this.positionX = position.x;
        this.positionY = position.y;
        utils.setCssElementPosition(element, position);
    }
    updateParentElementData() {
        this.parentRect = this.$element.parentElement.getBoundingClientRect();
        this.parentOffset = {
            left: this.$element.parentElement.offsetLeft,
            top: this.$element.parentElement.offsetTop
        };
    }
    onStart(event) {
        this.isDragging = true;
        this.$element.style.pointerEvents = 'none';
        this.$element.style.position = 'absolute';
        this.gridsterPrototype.dragItemStart(this, event);
        this.start.emit({ item: this.item });
    }
    onDrag(event) {
        this.gridsterPrototype.updatePrototypePosition(this, event);
    }
    onStop(event) {
        this.gridsterPrototype.dragItemStop(this, event);
        this.isDragging = false;
        this.$element.style.pointerEvents = 'auto';
        this.$element.style.position = '';
        utils.resetCSSElementPosition(this.$element);
        if (this.config.helper) {
            this.$element.parentNode.removeChild(this.$element);
        }
    }
    provideDragElement() {
        let dragElement = this.elementRef.nativeElement;
        if (this.config.helper) {
            dragElement = (dragElement).cloneNode(true);
            document.body.appendChild(this.fixStylesForBodyHelper(dragElement));
        }
        else {
            this.fixStylesForRelativeElement(dragElement);
        }
        return dragElement;
    }
    fixStylesForRelativeElement(el) {
        if (window.getComputedStyle(el).position === 'absolute') {
            return el;
        }
        const rect = this.elementRef.nativeElement.getBoundingClientRect();
        this.containerRectange = el.parentElement.getBoundingClientRect();
        el.style.position = 'absolute';
        this.setElementPosition(el, {
            x: rect.left - this.containerRectange.left,
            y: rect.top - this.containerRectange.top
        });
        return el;
    }
    /**
     * When element is cloned and append to body it should have position absolute and coords set by original
     * relative prototype element position.
     */
    fixStylesForBodyHelper(el) {
        const bodyRect = document.body.getBoundingClientRect();
        const rect = this.elementRef.nativeElement.getBoundingClientRect();
        el.style.position = 'absolute';
        this.setElementPosition(el, {
            x: rect.left - bodyRect.left,
            y: rect.top - bodyRect.top
        });
        return el;
    }
};
GridsterItemPrototypeDirective.ctorParameters = () => [
    { type: NgZone },
    { type: ElementRef },
    { type: GridsterPrototypeService }
];
__decorate([
    Output(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "drop", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "start", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "cancel", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "enter", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "out", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "data", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], GridsterItemPrototypeDirective.prototype, "config", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "w", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "wSm", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "wMd", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "wLg", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "wXl", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "h", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "hSm", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "hMd", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "hLg", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], GridsterItemPrototypeDirective.prototype, "hXl", void 0);
GridsterItemPrototypeDirective = __decorate([
    Directive({
        selector: '[ngxGridsterItemPrototype]'
    }),
    __metadata("design:paramtypes", [NgZone,
        ElementRef,
        GridsterPrototypeService])
], GridsterItemPrototypeDirective);
export { GridsterItemPrototypeDirective };
